### PR TITLE
Feature/cloud apps

### PR DESCRIPTION
**Pull request by Jskobos**
### To Test

Use dev services

### To Do
- [x] Fix all tests
- [x] Fix the custom label generator (done but can be more robust)
- [x] handle error states for if regions, linodes, or images failed to load
- [x] Move SSHKeys HOC to the container
- [x] ~~Move Checkout bar to the container (maybe we shouldn't do this)~~
- [x] Reset form state on tab change
- [x] Pass down UDF errors to UDF form and map the errors to the field
   * turns out, you can map a UDF error to a field because API errors return `{field: 'wp_pass', reason: 'reason goes here'}`
- [x] Refactor CASelectStackScriptPanel to not do any tabs handling. The panel is now tab-less
- [x] Fix Checkout bar positioning
- [ ] Find source of app crashing bug `cannot read property "render" of undefined`
- [ ] Confirm we can bookmark URLs that open the _Create from App_ flow with an app pre-selected
  * should we change the URL to something like `?type=one-click&appID=1234`?
- [x] Do null checking on Images, Types, and Linodes data
- [x] Pass down Redux loading and error states to their panels
- [x] Fix pagination code in the SelectLinodePanel
- [x] Deprecate use of SelectStackScriptPanel.tsx
   * it's only usage is in RebuildFromStackScript.tsx
- [x] Rename files to not include the _CA_ prefix
- [x] Add tests for all functions in the `formUtilities.ts` file
- [ ] Once Adam is done adding meta data to GET `/stackscripts`, source data from that endpoint
- [x] Destructure state and props in container component and pass down `{ ...rest }`
- [x] Remove "My StackScripts" and "Linode StackScripts" from StackScript Landing
- [x] Unify `filterUDFErrors` functions with RebuildFromStackScript and FromStackScriptContent
- [ ] GA Tracking 
- [x] Add tests for `LinodeCreate/utilities.ts`
- [ ] Fix regression - should be able to click a stackscript row to select it

### Bugs
-------
~~1. Click on My Images > My StackScripts~~
~~2. Click on One-Click~~
~~3. Crash - `cannot read property 'render' of undefined`~~

-------

### CODE REVIEW

File: src/features/StackScripts/stackScriptUtils.ts [hunk 1]
```diff
@@ -19,6 +19,17 @@ export const getStackScriptsByUser = (
...
   });
 
+/**
+ * helper function to get Cloud Apps StackScripts
+ *
+ * for the prototype, all the apps we need are going to be uploaded to
+ * Christine Puk's account. Keep in mind that the Linux distros will be missing from this
+ * list because we're intentionally not including the distros in this view
+ */
+export const getCloudAppsStackScripts = () => {
```
Comment (martinmckenna) [inline src/features/StackScripts/stackScriptUtils.ts]: @todo remove this

File: src/features/linodes/LinodesCreate/TabbedContent/formUtilities.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+};
+
+/**
+ * filter out all the UDF errors from our error state.
+ * To do this, we compare the keys from the error state to our "errorResources"
+ * map and return all the errors that don't match the keys in that object
+ *
+ * @todo test this function
+ */
+export const filterUDFErrors = (
```
Comment (martinmckenna) [inline src/features/linodes/LinodesCreate/TabbedContent/formUtilities.ts]: @todo test these both

**Comment by martinmckenna** [comment]
Screen Shot 2019-03-20 at 1 01 20 PM

While on the one-click view

**Comment by martinmckenna** [comment]
Screen Shot 2019-03-20 at 1 51 50 PM

fix dat

File: src/features/StackScripts/stackScriptUtils.ts (hunk 1 above)
Comment (martinmckenna) [reply src/features/StackScripts/stackScriptUtils.ts]: done

File: src/features/linodes/LinodesCreate/TabbedContent/formUtilities.ts (hunk 2 above)
Comment (martinmckenna) [reply src/features/linodes/LinodesCreate/TabbedContent/formUtilities.ts]: done

File: src/features/linodes/LinodesCreate/utilities.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
+import { compose, find, lensPath, prop, propEq, set } from 'ramda';
+import { displayType } from 'src/features/linodes/presentation';
+import { ExtendedType } from './SelectPlanPanel';
+import { ExtendedLinode } from './types';
+
+export const extendLinodes = (
```
Comment (johnwcallahan) [inline src/features/linodes/LinodesCreate/utilities.ts]: Could we write some documentation or tests for this function?

File: src/features/StackScripts/UserDefinedFieldsPanel/UserDefinedFieldsPanel.tsx [hunk 4]
```diff
@@ -40,82 +50,136 @@ interface Props {
...
-      {userDefinedFields!.map((field: Linode.StackScript.UserDefinedField) => {
-        return renderField(field);
-      })}
-    </Paper>
-  );
+
+        {/* Optional Fields */}
+        {optionalUDFs.length !== 0 && (
+          <ShowMoreExpansion
+            name="Show Advanced Options"
```
[the remark this answers is not in the document]
Comment (martinmckenna) [reply src/features/StackScripts/UserDefinedFieldsPanel/UserDefinedFieldsPanel.tsx]: yep - This UX is changing a bit, but will keep in mind @WilkinsKa1

File: src/features/StackScripts/SelectStackScriptPanel/CASelectStackScriptPanel.tsx [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+                currentFilterType={null}
+                isSelecting={true}
+              />
+              <tbody>
+                <StackScriptSelectionRow
+                  key={stackScript.id}
+                  label={stackScript.label}
+                  stackScriptUsername={stackScript.username}
+                  disabledCheckedSelect
+                  onSelect={() => {}}
```
[the remark this answers is not in the document]
Comment (martinmckenna) [reply src/features/StackScripts/SelectStackScriptPanel/CASelectStackScriptPanel.tsx]: Yeah. This is when your stackscript is preselected based on the query param. We should add documentation here to explain

File: src/features/StackScripts/SelectStackScriptPanel/CASelectStackScriptPanel.tsx [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+            className={classes.header}
+            role="header"
+            variant="h2"
+            data-qa-tp-title
+          >
+            {header}
+          </Typography>
+          {stackScriptError && (
+            <Typography variant="body2">
+              An error occured while loading the selected StackScript.
```
[the remark this answers is not in the document]
Comment (Jskobos) [reply src/features/StackScripts/SelectStackScriptPanel/CASelectStackScriptPanel.tsx]: Fixed this in Marty's PR (which moves/renames this file)

File: src/features/linodes/LinodesCreate/LinodeCreateContainer.tsx [hunk 7]
```diff
@@ -0,0 +1,655 @@
...
+      .then(response => {
+        this.setState({
+          appInstancesLoading: false,
+          appInstances: response.data
+        });
+      })
+      .catch(e => {
+        this.setState({
+          appInstancesLoading: false,
+          appInstancesError: 'There was an error loading Cloud Apps.'
```
[the remark this answers is not in the document]
Comment (Jskobos) [reply src/features/linodes/LinodesCreate/LinodeCreateContainer.tsx]: Fixed

File: src/features/linodes/LinodesCreate/PublicImages.tsx [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+const PublicImages: React.StatelessComponent<CombinedProps> = props => {
+  const {
+    classes,
+    disabled,
+    images,
+    handleSelection,
+    oldImages,
+    selectedImageID
+  } = props;
+  const renderImages = (images: Linode.Image[]) =>
```
[the remark this answers is not in the document]
Comment (Jskobos) [reply src/features/linodes/LinodesCreate/PublicImages.tsx]: Fixed

File: src/features/linodes/LinodesCreate/PublicImages.tsx [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+  images: Linode.Image[];
+  oldImages: Linode.Image[];
+  selectedImageID?: string;
+  disabled?: boolean;
+  handleSelection: (id: string) => void;
+}
+
+type CombinedProps = Props & WithStyles<ClassNames>;
+
+const distroIcons = {
```
[the remark this answers is not in the document]
Comment (Jskobos) [reply src/features/linodes/LinodesCreate/PublicImages.tsx]: Added

File: src/features/linodes/LinodesCreate/utilities.ts (hunk 3 above)
Comment (Jskobos) [reply src/features/linodes/LinodesCreate/utilities.ts]: Lol, this is an old Rob function and I don't really understand what it's doing, but I'll try

File: src/features/linodes/LinodesCreate/TabbedContent/FromBackupsContent.tsx [hunk 10]
```diff
@@ -126,20 +84,11 @@ const filterLinodesWithBackups = (linodes: Linode.LinodeWithBackups[]) => {
 export class FromBackupsContent extends React.Component<CombinedProps, State> {
```
[the remark this answers is not in the document]
Comment (Jskobos) [reply src/features/linodes/LinodesCreate/TabbedContent/FromBackupsContent.tsx]: Fixed

File: src/features/StackScripts/SelectStackScriptPanel/CASelectStackScriptPanel.tsx [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+  RenderGuardProps &
+  WithStyles<ClassNames>;
+
+interface State {
+  stackScript?: Linode.StackScript.Response;
+  stackScriptError: boolean;
+  stackScriptLoading: boolean;
+}
+
+class SelectStackScriptPanel extends React.Component<CombinedProps, State> {
```
[the remark this answers is not in the document]
Comment (martinmckenna) [reply src/features/StackScripts/SelectStackScriptPanel/CASelectStackScriptPanel.tsx]: fixed in #4709

File: src/features/linodes/LinodesCreate/SelectAppPanel.tsx [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+      </Panel>
+    );
+  }
+
+  if (!appInstances) {
+    return null;
+  }
+
+  /** hacky and bad */
+  const interceptedError = error ? 'You must select an App to create from' : '';
```
[the remark this answers is not in the document]
Comment (martinmckenna) [reply src/features/linodes/LinodesCreate/SelectAppPanel.tsx]: fixed in #4709

File: src/features/StackScripts/SelectStackScriptPanel/CASelectStackScriptPanel.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (martinmckenna) [reply src/features/StackScripts/SelectStackScriptPanel/CASelectStackScriptPanel.tsx]: fixed in #4721

File: src/features/linodes/LinodesCreate/utilities.ts (hunk 3 above)
Comment (martinmckenna) [reply src/features/linodes/LinodesCreate/utilities.ts]: fixed #4726

File: src/components/ShowMoreExpansion/ShowMoreExpansion.tsx [hunk 13]
```diff
@@ -42,25 +42,33 @@ const styles: StyleRulesCallback = theme => ({
 
 interface Props {
   name: string;
+  defaultExpanded?: boolean;
 }
 
 interface State {
-  open: boolean;
+  open: boolean | undefined;
```
Comment (rulikrulit) [inline src/components/ShowMoreExpansion/ShowMoreExpansion.tsx]: Can also be simplified with `open?: boolean;`. I'm fine with any but just to keep uniformity

File: src/features/StackScripts/UserDefinedFieldsPanel/UserDefinedFieldsPanel.tsx [hunk 14]
```diff
@@ -40,82 +50,136 @@ interface Props {
...
+        <Grid item xs={12} key={field.name}>
+          <UserDefinedSelect
+            field={field}
+            updateFormState={handleChange}
+            udf_data={udf_data}
+            updateFor={[udf_data[field.name], error]}
+            isOptional={isOptional}
+            key={field.name}
+            error={error}
+          />{' '}
```
Comment (rulikrulit) [inline src/features/StackScripts/UserDefinedFieldsPanel/UserDefinedFieldsPanel.tsx]: We'd better use &nbsp; HTML symbol here if we need a space

File: src/features/StackScripts/stackScriptUtils.ts [hunk 15]
```diff
@@ -150,11 +134,29 @@ export const generateCatchAllFilter = (searchTerm: string) => {
...
+) => {
+  let type;
+  let subtype;
+  switch (username) {
+    case 'linode':
+      // This is a Cloud App (unless it isn't, which we are unable to handle at this time)
+      type = 'One-Click';
+      subtype = 'One-Click%20Apps';
+      break;
+    case currentUser:
```
Comment (rulikrulit) [inline src/features/StackScripts/stackScriptUtils.ts]: Should this case include account stackScripts too?

Comment (martinmckenna) [reply src/features/StackScripts/stackScriptUtils.ts]: nah - the tab and URL is just called "My StackScripts"

To your point though, we probably should get an answer from stakeholders on whether we're calling them "Account StackScripts" or "My StackScripts."

IMO, "My" doesn't imply all the StackScripts on the account

**Pull request merged**